yt_dlp/downloader/mhtml.py

```diff
@@ -0,0 +1,175 @@
+
+
+class MhtmlFD(FragmentFD):
+    FD_NAME = 'mhtml'
+    _EXTENSION = 'mhtml'
+
+    _STYLESHEET = """\
+html, body {
+    margin: 0;
+    padding: 0;
+    height: 100vh;
+}
+
+html {
+    overflow-y: scroll;
+    scroll-snap-type: y mandatory;
+}
+
+body {
+    scroll-snap-type: y mandatory;
+    display: flex;
+    flex-flow: column;
+}
+
+body > figure {
+    max-width: 100vw;
+    max-height: 100vh;
+    scroll-snap-align: center;
+}
+
+body > figure > figcaption {
+    text-align: center;
+    height: 2.5em;
+}
+
+body > figure > img {
+    display: block;
+    margin: auto;
+    max-width: 100%;
+    max-height: calc(100vh - 5em);
+}
+"""
+
+    def _gen_cid(self, i, fragment, frag_boundary):
+        return '%u.%s@yt-dlp.github.io.invalid' % (i, frag_boundary.decode('us-ascii'))
+
+    def _gen_stub(self, fragments, frag_boundary):
+        output = io.StringIO()
+
+        output.write(
+            '<!DOCTYPE html>'
+            '<html>'
+            '<head>'
+            ''  '<meta name="generator" content="yt-dlp %s">'
+            ''  '<title>yt-dlp downloaded slideshow</title>'
+            ''  '<style>%s</style>'
+            '<body>'
+            '\n'
+            % (
+                YT_DLP_VERSION,
+                self._STYLESHEET,
+            )
+        )
+
+        t0 = 0
+        for i, frag in enumerate(fragments):
+            t1 = t0 + frag['duration']
+            output.write(
+                '<figure>'
+                    '<figcaption>Slide #%u: %s – %s (duration: %s)</figcaption>'
+                    '<img src="cid:%s">'
+                '</figure>\n'
+                % (
+                    i + 1,
+                    srt_subtitles_timecode(t0),
+                    srt_subtitles_timecode(t1),
+                    formatSeconds(frag['duration']),
+                    self._gen_cid(i, frag, frag_boundary),
+                )
+            )
+            t0 = t1
+
+        return output.getvalue()
+
+    def real_download(self, filename, info_dict):
+        fragment_base_url = info_dict.get('fragment_base_url')
+        fragments = info_dict['fragments'][:1] if self.params.get(
+            'test', False) else info_dict['fragments']
+
+        ctx = {
+            'filename': filename,
+            'total_frags': len(fragments),
+        }
+
+        self._prepare_and_start_frag_download(ctx)
+
+        extra_state = ctx.setdefault('extra_state', {
+            'header_written': False,
+            'mime_boundary': str(uuid.uuid4()).replace('-', ''),
+        })
+
+        frag_boundary = extra_state['mime_boundary'].encode('us-ascii')
+
+        if not extra_state['header_written']:
+            stub = self._gen_stub(fragments, frag_boundary)
+
+            ctx['dest_stream'].write(
+                b'MIME-Version: 1.0\r\n'
+                b'From: <nowhere@yt-dlp.github.io.invalid>\r\n'
+                b'To: <nowhere@yt-dlp.github.io.invalid>\r\n'
+                b'Subject: yt-dlp downloaded slideshow\r\n'
```

Comment:
This may help content-type sniffing: first line starting with `From:` may identify the file as a MIME message, which, in fact, this is. As I remember, Chromium also leaves a dummy `From` header in the file. And it’s otherwise harmless anyway. I see no reason to risk incompatibility by omitting it.

Reply:
I am not a fan of using these fake "email ids" in this. But I guess you are right. My personal dislike is not really reason enough to deviate from the specs